### ISSUE TITLE
💡 [FEAT] Highscore Page

### ISSUE THREAD

**Issue by a-usr** [kraanzu/smassh#79]
### Is your feature request related to a problem? 😢 Please describe.

No

### Describe the solution you'd like 🤔

Basically a Tab that displays a Highscore for Words per Minute, and the Accuracy you had when you set the Highscore. That Highscore would of course have to be saved somewhere, so I propose just pickling the number and then saving it in binary in a file next to the config. This has the advantage of making tampering with the highscore more difficult.

### Additional context 📝

Would be nice for the Unixporn community to have. And would thus increase popularity of your project. Also, on a side note, your project got starred by Elkowar, so congrats!

**Comment by kraanzu** [comment]
Hi @a-usr !

> That Highscore would of course have to be saved somewhere

Yes all the info are being saved into `~/.local/share/smassh/smassh.json` or if you're not on linux, they are saved at:

```
python -c "import appdirs; print(appdirs.user_data_dir('smassh'))"
```

> binary in a file next to the config. This has the advantage of making tampering with the highscore more difficult.

I don't think I can stop tampering even with a binary file but I'll check if it has any effect on the performance

------

Additional points:
You should see `crown` icon if you score more than your highscore

![image](https://github.com/kraanzu/smassh/assets/97718086/fd6be6f7-3c7e-48fe-99f3-0174bfd9df54)

I'm thinking about implementing a user dashboard just like we have in monkeytype but there's currently not a good support for graphs in the underlying TUI lib so I might have to come up with something myself.

I think I can at least currently implement these parts:


![image](https://github.com/kraanzu/smassh/assets/97718086/2546258a-8beb-4d12-a9a7-1aae37e0485c)

![image](https://github.com/kraanzu/smassh/assets/97718086/c03aeaff-116b-4bc4-a1e3-2a6a6dde0025)

**Comment by a-usr** [comment]
Hey @kraanzu 
> Yes all the info are being saved into `~/.local/share/smassh/smassh.json`

May I ask why specifically this directory? Why not `~/.config/smassh` ?

> I don't think I can stop tampering even with a binary file but I'll check if it has any effect on the performance

My point was to make it harder to fake a Highscore, as users without programming experience probably wouldn't go through the trouble of figuring out how to edit that specific binary file correctly. For most normal users, just mentioning the words "binary" and "edit" in one sentence is very scary

> 
> Additional points:
> You should see `crown` icon if you score more than your highscore
> 

Yeah, good Idea!


> I'm thinking about implementing a user dashboard just like we have in monkeytype but there's currently not a good support for graphs in the underlying TUI lib so I might have to come up with something myself.
> 
> 

Where would you host the database for that?
Also, MonkeyType doesn't have to deal with this since it's a website, but what would happen if the user was offline? And how do you detect when a person is trying to upload a tampered value? Don't get me wrong, I'm all for this Idea, and would even help realising it, but there are some considerations to be made.

**Comment by kraanzu** [comment]
> May I ask why specifically this directory? Why not ~/.config/smassh ?

I'm using [this library](https://github.com/ActiveState/appdirs) to store config/data.
So there are 2 funcs that I'm using  `user_config_dir` and `user_data_dir` 

This also aligns with [XDG directory specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html)

I also wasn't aware of this and stored everything in the .config dir before haha. I learned about this from https://github.com/kraanzu/dooit/issues/12 and https://github.com/kraanzu/smassh/issues/48

> For most normal users, just mentioning the words "binary" and "edit" in one sentence is very scary

Yes, I understand. If I take this thing online then it can become a little bit more vulnerable

**Comment by a-usr** [comment]
> Yes, I understand. If I take this thing online then it can become a little bit more vulnerable

Feels like you replied to the pre-edit version of my comment. I actually mistook this screenshot for a global leaderboard:
![image](https://github.com/kraanzu/smassh/assets/81042605/138b5c16-bf28-4538-99d0-7a8677f910ba)

That's why I assumed you wanted to create Webservices for this project.

**Comment by kraanzu** [comment]
> should you plan on implementing those

This I'll implement after I'm satisfied with the local client. Make sure it has enough features and then I'll use something like appwrite to store the global results

**Comment by a-usr** [comment]
> also I think the XDG specs are clear now for you? :)

Yes, after taking a look at #48 I realized that unlike what I thought the config and user data are stored separately from eachother on Linux. The root of this misconception was that I had only used smassh on windows, not on Linux.
